prow/tide/queryutil/queryutil.go

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package queryutil provides a utility that helps to query api v4 with
+// given number of goroutines running concurrently
+package queryutil
+
+import (
+	"sync"
+)
+
+type QueryData struct {
+	Org         string
+	Query       string
+	MetricIndex int
+}
+
+// QueriesInParallel runs process function concurrently respecting maximum number of goroutines
+func QueriesInParallel(goroutines uint, queries []QueryData, process func(QueryData)) {
```

Comment:
Sorry, but now this construct is awkward. The reason that the previous code was not ideal was that it was doing the same thing - running GraphQL queries using a client against the same server, etc. This incarnation is a "generic" thread pool where the arguments are passed in a slice that's coupled to the execution functions being called. I don't think there's a huge amount of utility in "generic" solutions until the type system supports parametric polymorphism. To wit - this helper could be even more utilitarian if you close over the arguments to the function itself, but, again, that's not really that interesting of a module:

```go
func DoWhateverInParallel(concurrency uint, work []func()) {}
```

Notably, the actual sending of the graphQL query and the incrementation of the error counters is now (still) duplicated in both callers, and *that* was the common thread which could have used "DRY". Consider the case in the future when someone wants to change how `tide` interacts with GitHub v4 API. Given this current code, they'd need to go make the same change in a number of places.

Reply:
Well, every query is a specific thing and you want a struct for its representation. Unless you decide to compose it dynamically (use reflect) you can't unify different cases. In `blockers` you have issues in `tide` you have PRs. Max you can do is this + maybe some kind of error handling and metrics (see another comment). Any way you will end up with the same quite generic mechanism.